MH-13510 Fix conflict detection for non-admin users and for multiple events

smarquard opened the pull request:
The scheduler service and Admin UI endpoints need to be able to get conflicting
events so as to return conflict information, even if the user attempting to
schedule a conflicting event does not have access to the events conflicted with.

https://opencast.jira.com/browse/MH-13510 for steps to reproduce.

smarquard committed:
MH-13510 Fix conflict detection for non-admin users

The scheduler service and Admin UI endpoints need to be able to get conflicting
events so as to return conflict information, even if the user attempting to
schedule a conflicting event does not have access to the events conflicted with.

smarquard commented:
There may be a similar issue in the External API endpoints - haven't checked that yet.

smarquard commented:
May also need some tweaks to the exception handling here.

gregorydlogan commented:
@smarquard Do you have a timeline for looking at the external API components?  Or should this be merged as is?

gregorydlogan commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:
```diff
@@ -2051,19 +2055,29 @@ public Response getNewConflicts(@FormParam("metadata") String metadata) throws N
+            if (StringUtils.isNotEmpty(eventId) && eventId.equals(e.getIdentifier())) {
+              continue;
+            }
+            eventsJSON.add(convertEventToConflictingObject(e.getTechnicalStartTime(), e.getTechnicalEndTime(), e.getTitle()));
```
Doesn't this leak the event titles?  If you don't have access to these events, shouldn't you not know their titles?

smarquard replied:
Yes, it does return event title.

gregorydlogan commented on modules/scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java:
```diff
@@ -1081,8 +1081,21 @@ public int getEventCount() throws SchedulerException {
+      SecurityUtil.runAs(securityService, organization, user, () -> {
+        try {
+          conflictingEvents.addAll(persistence.getEvents(captureDeviceID, startDate, endDate, Util.EVENT_MINIMUM_SEPARATION_MILLISECONDS)
+            .parallelStream().map(this::getEventMediaPackage).collect(Collectors.toList()));
```
Similar to above, this leaks the mediapackages, doesn't it?

smarquard replied:
Seems like it would through the /recordings/conflicts* REST endpoints. Needs some more work here as I'm currently getting a 500 error on a conflict check because of a user/org issue.

gregorydlogan reviewed (commented):
To be clear, I don't particularly care if we're leaking the event titles, or even their full mediapackages, but I recall someone objecting to this at one point.

gregorydlogan commented:
Assigning to @lkiesow so that he remembers to ask his contacts at SWITCH if this privacy is still important.

gregorydlogan commented:
CC @staubesv, we think it was SWITCH that wanted this privacy, but we're not sure :D

smarquard commented:
The json response in the AdminUI endpoint for a conflict check returns the start and end times of the conflicting event and no other info (e.g. who created the event, presenter or title).

[{"start":"2019-12-21T10:00:00Z","end":"2019-12-21T10:55:00Z","title":"super secret event even the title is eyes-only"}]

That appears to be by design.

I think some sites might consider it useful to also show who created the conflicting event. But I'd say changing that (or removing the title) is out-of-scope for this PR; the goal is just to make it work properly as designed.

I haven't checked how the external API handles this, and won't investigate that for this PR.

smarquard commented:
For the /recordings/conflicts* endpoints, what would people think about making these admin-only, e.g. restricted via security config to an appropriate role, or return unauthorized if you're not admin.

These are "grey" APIs since they're not used by the UI and are also not guaranteed stable, so that seems a reasonable compromise. Alternately we need to add more code to make their behaviour consistent with the admin UI endpoint (and external API which I haven't checked yet). But as they're effectively deprecated, that seems a waste.

gregorydlogan commented:
That seems reasonable to me.  Perhaps bringing this up on list for discussion here would make sense.

lkiesow commented:
I've contacted Matthias at SWITCH as you asked. Let's see what his take on this is.

---

Apart from that, instead of the user switching to get maximum privileges and then requesting all data, wouldn't it suffice to just add an empty response if the user should not have access? After all, we do already know that the events are conflicting, don't we?

Something like this:

```java
  private List<JValue> convertToConflictObjects(final String eventId, final List<MediaPackage> events) throws SearchIndexException {
    final List<JValue> eventsJSON = new ArrayList<>();
    for (final MediaPackage event : events) {
      final Opt<Event> eventOpt = getIndexService().getEvent(event.getIdentifier().compact(), getIndex());
      if (eventOpt.isSome()) {
        final Event e = eventOpt.get();
        if (StringUtils.isNotEmpty(eventId) && eventId.equals(e.getIdentifier())) {
          continue;
        }
        eventsJSON.add(convertEventToConflictingObject(e.getTechnicalStartTime(), e.getTechnicalEndTime(), e.getTitle()));
      } else {
        eventsJSON.add(obj(
          f("start", null),
          f("end", null),
          f("title", null))); // or  f("title", v("<hidden>"))
      }
    }
    return eventsJSON;
  }
```

smarquard commented:
Two followups to the above:

First, we don't need to worry about /recordings/conflicts leaking mediapackages because that endpoint is already restricted:

    <sec:intercept-url pattern="/recordings/**" method="GET" access="ROLE_ADMIN, ROLE_CAPTURE_AGENT" />

Second, I don't like the idea of returning an empty response if the user doesn't have access, because ideally we would like users to solve conflict issues themselves (if possible), without requesting support from someone with higher access levels. For example, a common scenario we have found is that people will try to schedule say 10am - 11am when there's another recording starting at 11am. If they know the conflict is with that event, then they can adjust their own recording to end at 10:55 rather than just getting a generic "conflict" failure which doesn't suggest what to do.

For the same reason I can see the value of letting the user know who scheduled the conflicting event (possibly more helpful than the title). But either way, the goal of this PR is just to make the as-designed functionality work and I think changes to what info is returned should be in a new PR.

smarquard committed:
MH-13510 Use the default organization for conflict checks

smarquard commented:
I think this is OK to merge as-is.

I looked at the external API behaviour briefly. If you POST to /api/events with the system digest username/password and there's a conflict, you get back a response that you'd expect:

````
409 Conflict
[{"start":"2020-01-02T05:30:00Z","end":"2020-01-02T06:30:00Z","title":"2020-01-02 07:30"}]
````

If you POST to /api/events as an unprivileged user with ROLE_API_* roles (I had to add more than expected before getting past the 403 error), it fails with a 500 internal server error that seems an unrelated issue (a folder in workspace already exists). That seems like a separate investigation.

smarquard commented:
ping @lkiesow, anything else you want to do here?

lkiesow commented:
True, we don't really need to worry about the scheduler endpoint. `ROLE_ADMIN` has access anyway and [according to our docs, a capture agent wouldn't access this](https://docs.opencast.org/develop/developer/#modules/capture-agent/capture-agent/).

> Second, I don't like the idea of returning an empty response if the user doesn't have access, because ideally we would like users to solve conflict issues themselves

In that case, why don't you give users read access to each other's recordings? That would make them show up in the conflict handling with title even with the proposed change.

---

> For example, a common scenario we have found is that people will try to schedule say 10am - 11am when there's another recording starting at 11am.

If a conflict shows up for that, it's a clear bug. I remember that we discussed that start and end for recordings should be conformant to the [iCal definition of start and end](https://tools.ietf.org/html/rfc5545#section-3.6.1) (we are convertig those to iCal after all) which states:

>  The "DTSTART" property for a "VEVENT" specifies the inclusive start of the event.

and

>  The "DTEND" property for a "VEVENT" calendar component specifies the non-inclusive end of the event.

Hence, two events (9am to 10am) and (10am to 11am) would not conflict since 10am alrearly belongs to the second event and is not part of the first event.

smarquard commented:
> In that case, why don't you give users read access to each other's recordings? That would make them show up in the conflict handling with title even with the proposed change.

That's not really a reasonable production requirement. Typically the conflicting recordings are in different series with totally different ACLs and permissions. 

The edge cases above are also not really core to this issue because the conflict could be a larger overlap (or for the same time) because of some confusion like venue changes or similar.

smarquard commented:
I'm restating here that this is a bugfix: Opencast at present, by design, intends to return start, end and title for conflicting events (hence the existence of convertEventToConflictingObject). This change just makes it work as-designed.

Functional changes about what is returned (no info, or some other info, or all info) should be in another issue.

smarquard commented:
For reference, the PR relating to adjacent events is #1370

smarquard commented:
ping @lkiesow @gregorydlogan I'd like to get this wrapped up. I don't believe there's anything further to do in this PR.

smarquard commented:
Actually we seem to have a problem locally here under some circumstances:

```
 SecurityUtil.runAs(securityService, organization, user, () -> {
        try {
          conflictingEvents.addAll(persistence.getEvents(captureDeviceID, startDate, endDate, Util.EVENT_MINIMUM_SEPARATION_MILLISECONDS)
            .parallelStream().map(this::getEventMediaPackage).collect(Collectors.toList()));
```

where we're getting an exception from one of the threads created by the parallelStream():

```
Caused by: java.lang.IllegalStateException: No organization is set in security context
	at org.opencastproject.kernel.security.SecurityServiceSpringImpl.getUser(SecurityServiceSpringImpl.java:94) ~[?:?]
	at org.opencastproject.assetmanager.impl.AssetManagerWithSecurity.isAdmin(AssetManagerWithSecurity.java:229) ~[?:?]
	at org.opencastproject.assetmanager.impl.AssetManagerWithSecurity.access$000(AssetManagerWithSecurity.java:61) ~[?:?]
	at org.opencastproject.assetmanager.impl.AssetManagerWithSecurity$1.select(AssetManagerWithSecurity.java:138) ~[?:?]
	at org.opencastproject.scheduler.impl.SchedulerServiceImpl.getEventMediaPackage(SchedulerServiceImpl.java:1526) ~[?:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[?:?]
	at java.util.Vector$VectorSpliterator.forEachRemaining(Vector.java:1440) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) ~[?:?]
	at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:747) ~[?:?]
	at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:721) ~[?:?]
	at java.util.stream.AbstractTask.compute(AbstractTask.java:327) ~[?:?]
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731) ~[?:?]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[?:?]
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) ~[?:?]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) ~[?:?]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) ~[?:?]
```

so looks like the approach used here in unsafe in some way, as the SecurityUtil.runAs() method is setting the user/organization in threadlocal for the current thread, but getEventMediaPackage is called in another thread which doesn't have a user/organization set.

smarquard committed:
MH-13510 Use stream() rather than parallelStream()

to avoid threadLocal issues where the organization and user is not
set in the threads spawned for parallel processing.

gregorydlogan commented:
@smarquard there should be more stack trace, or another one immediately before/after which ought to tell you what called `getEventMediaPackage` - it's not like it calls itself, nor is it the REST endpoint directly.  Grepping for that method, I see calls from the admin ui, and scheduler-impl, so it's at least a relatively narrow field.

smarquard commented:
It's called from here:

 conflictingEvents.addAll(persistence.getEvents(captureDeviceID, startDate, endDate, Util.EVENT_MINIMUM_SEPARATION_MILLISECONDS)
            .parallelStream().map(this::getEventMediaPackage).collect(Collectors.toList()));

Replacing the parallelStream() with stream() fixes the issue (https://github.com/cilt-uct/opencast/commit/e4df0ed7131eb302cec46d5c79f505e10dd0bfa9), and shouldn't be a serious performance concern as in most cases there won't be a lot of conflicting events.

Alternately we'd have to put another runAs inside the parallelStream() or a wrapper around getEventMediaPackage or some way of dealing with threads spawned by parallelStream to make them inherit the security context.

Interesting problem, but I'm happy with the solution in the last commit.

smarquard commented:
For reference, the full stack trace is below.

Prior to the fix, this was reproducible by creating three events in a venue at 9am, 10am, 11am and then trying to create an event that spanned them all (say 7am to 12pm).

```
2020-02-21 13:30:34,887 ERROR o.o.a.e.AbstractEventEndpoint [qtp630545973-356783] Unable to find conflicting events for beattielt, 2020-02-25T12:00:00Z, 2020-02-25T12:55:00Z
org.opencastproject.scheduler.api.SchedulerException: java.lang.IllegalStateException: java.lang.IllegalStateException: No organization is set in security context
	at org.opencastproject.scheduler.impl.SchedulerServiceImpl.findConflictingEvents(SchedulerServiceImpl.java:1115) ~[?:?]
	at org.opencastproject.adminui.endpoint.AbstractEventEndpoint.getNewConflicts(AbstractEventEndpoint.java:2006) [40:opencast-admin-ui:8.0.0.SNAPSHOT]
	at sun.reflect.GeneratedMethodAccessor145.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179) [246:org.apache.cxf.cxf-core:3.2.6]
	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96) [246:org.apache.cxf.cxf-core:3.2.6]
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:193) [247:org.apache.cxf.cxf-rt-frontend-jaxrs:3.2.6]
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:103) [247:org.apache.cxf.cxf-rt-frontend-jaxrs:3.2.6]
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59) [246:org.apache.cxf.cxf-core:3.2.6]
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96) [246:org.apache.cxf.cxf-core:3.2.6]
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308) [246:org.apache.cxf.cxf-core:3.2.6]
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121) [246:org.apache.cxf.cxf-core:3.2.6]
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:267) [254:org.apache.cxf.cxf-rt-transports-http:3.2.6]
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234) [254:org.apache.cxf.cxf-rt-transports-http:3.2.6]
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208) [254:org.apache.cxf.cxf-rt-transports-http:3.2.6]
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160) [254:org.apache.cxf.cxf-rt-transports-http:3.2.6]
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:216) [254:org.apache.cxf.cxf-rt-transports-http:3.2.6]
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:301) [254:org.apache.cxf.cxf-rt-transports-http:3.2.6]
	at org.opencastproject.kernel.rest.RestPublisher$RestServlet.handleRequest(RestPublisher.java:559) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:220) [254:org.apache.cxf.cxf-rt-transports-http:3.2.6]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707) [35:javax.servlet-api:3.1.0]
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:276) [254:org.apache.cxf.cxf-rt-transports-http:3.2.6]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1655) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.opencastproject.kernel.rest.JsonpFilter.doFilter(JsonpFilter.java:113) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.opencastproject.kernel.filter.proxy.TransparentProxyFilter.doFilter(TransparentProxyFilter.java:63) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.opencastproject.security.urlsigning.filter.UrlSigningFilter.doFilter(UrlSigningFilter.java:86) [192:opencast-urlsigning-verifier-service-impl:8.0.0.SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.opencastproject.kernel.security.RemoteUserAndOrganizationFilter.doFilter(RemoteUserAndOrganizationFilter.java:247) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.opencastproject.kernel.rest.CurrentJobFilter.doFilter(CurrentJobFilter.java:87) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:118) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:84) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.opencastproject.kernel.security.AsyncTimeoutRedirectFilter.doFilter(AsyncTimeoutRedirectFilter.java:60) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:113) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:103) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:113) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:146) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:54) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:45) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:98) [377:org.springframework.web:3.1.4.RELEASE]
	at org.springframework.security.oauth.provider.filter.OAuthProviderProcessingFilter.doFilter(OAuthProviderProcessingFilter.java:182) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:102) [377:org.springframework.web:3.1.4.RELEASE]
	at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilter(BasicAuthenticationFilter.java:150) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:102) [377:org.springframework.web:3.1.4.RELEASE]
	at org.springframework.security.web.authentication.www.DigestAuthenticationFilter.doFilter(DigestAuthenticationFilter.java:115) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:102) [377:org.springframework.web:3.1.4.RELEASE]
	at org.springframework.web.filter.CompositeFilter.doFilter(CompositeFilter.java:82) [377:org.springframework.web:3.1.4.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.authentication.ui.DefaultLoginPageGeneratingFilter.doFilter(DefaultLoginPageGeneratingFilter.java:91) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:183) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:105) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:87) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:192) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:160) [375:org.springframework.security.web:3.1.7.RELEASE]
	at org.opencastproject.kernel.security.SecurityFilter.doFilter(SecurityFilter.java:124) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.opencastproject.kernel.security.OrganizationFilter.doFilter(OrganizationFilter.java:135) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.opencastproject.kernel.filter.https.HttpsFilter.doFilter(HttpsFilter.java:73) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.opencastproject.kernel.rest.CleanSessionsFilter.doFilter(CleanSessionsFilter.java:89) [117:opencast-kernel:8.0.0.SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1634) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:215) [331:org.eclipse.jetty.websocket.server:9.4.12.v20180830]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71) [356:org.ops4j.pax.web.pax-web-jetty:7.2.5]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) [318:org.eclipse.jetty.security:9.4.12.v20180830]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1340) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:293) [356:org.ops4j.pax.web.pax-web-jetty:7.2.5]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473) [321:org.eclipse.jetty.servlet:9.4.12.v20180830]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1242) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80) [356:org.ops4j.pax.web.pax-web-jetty:7.2.5]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.server.Server.handle(Server.java:503) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:364) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260) [320:org.eclipse.jetty.server:9.4.12.v20180830]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305) [312:org.eclipse.jetty.io:9.4.12.v20180830]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) [312:org.eclipse.jetty.io:9.4.12.v20180830]
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118) [312:org.eclipse.jetty.io:9.4.12.v20180830]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333) [323:org.eclipse.jetty.util:9.4.12.v20180830]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310) [323:org.eclipse.jetty.util:9.4.12.v20180830]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168) [323:org.eclipse.jetty.util:9.4.12.v20180830]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126) [323:org.eclipse.jetty.util:9.4.12.v20180830]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366) [323:org.eclipse.jetty.util:9.4.12.v20180830]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765) [323:org.eclipse.jetty.util:9.4.12.v20180830]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683) [323:org.eclipse.jetty.util:9.4.12.v20180830]
	at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.lang.IllegalStateException: java.lang.IllegalStateException: No organization is set in security context
	at sun.reflect.GeneratedConstructorAccessor1253.newInstance(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:?]
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593) ~[?:?]
	at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677) ~[?:?]
	at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:735) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:714) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566) ~[?:?]
	at org.opencastproject.scheduler.impl.SchedulerServiceImpl.lambda$findConflictingEvents$2(SchedulerServiceImpl.java:1106) ~[?:?]
	at org.opencastproject.security.util.SecurityUtil.runAs(SecurityUtil.java:83) ~[?:?]
	at org.opencastproject.scheduler.impl.SchedulerServiceImpl.findConflictingEvents(SchedulerServiceImpl.java:1103) ~[?:?]
	... 113 more
Caused by: java.lang.IllegalStateException: No organization is set in security context
	at org.opencastproject.kernel.security.SecurityServiceSpringImpl.getUser(SecurityServiceSpringImpl.java:94) ~[?:?]
	at org.opencastproject.assetmanager.impl.AssetManagerWithSecurity.isAdmin(AssetManagerWithSecurity.java:229) ~[?:?]
	at org.opencastproject.assetmanager.impl.AssetManagerWithSecurity.access$000(AssetManagerWithSecurity.java:61) ~[?:?]
	at org.opencastproject.assetmanager.impl.AssetManagerWithSecurity$1.select(AssetManagerWithSecurity.java:138) ~[?:?]
	at org.opencastproject.scheduler.impl.SchedulerServiceImpl.getEventMediaPackage(SchedulerServiceImpl.java:1526) ~[?:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[?:?]
	at java.util.Vector$VectorSpliterator.forEachRemaining(Vector.java:1440) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) ~[?:?]
	at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:747) ~[?:?]
	at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:721) ~[?:?]
	at java.util.stream.AbstractTask.compute(AbstractTask.java:327) ~[?:?]
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731) ~[?:?]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[?:?]
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) ~[?:?]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) ~[?:?]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) ~[?:?]

```

smarquard commented:
This issue with the "No organization is set in security context" applies to conflict checks for admin users as well, and is reproducible on stable.opencast.org. So the conflict check is generally broken right now in 8.x if there's > 1 conflicting event, not only broken for non-admin users.

smarquard commented:
Can we wrap this up?

gregorydlogan commented:
I'm happy with this.  Merged.

KatrinIhler commented on modules/scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java:
```diff
@@ -1081,7 +1081,7 @@ public int getEventCount() throws SchedulerException {
       throws SchedulerException {
     try {
-      final Organization organization = securityService.getOrganization();
+      final Organization organization = new DefaultOrganization();
```
What was the reason for this change? It's possible that this might be breaking conflict checks for multitenant systems (as this organization is later used to look for conflicting events in the database).
